### PR TITLE
Fix: Lambda layer zip file exceeded deployment package size limit (#3329)

### PR DESCRIPTION
**Pull request by jessebrennan**
https://github.com/DataBiosphere/azul/issues/3329

Author

- [x] PR title references issue
- [x] Title of main commit references issue
- [x] PR is connected to Zenhub issue and description links to issue

Author (reindex)

- [x] Added `r` tag to commit title                         <sub>or this PR does not require reindexing</sub>
- [x] Added `reindex` label to PR                           <sub>or this PR does not require reindexing</sub>

Author (freebies & chains)

- [x] Freebies are blocked on this PR                       <sub>or there are no freebies in this PR</sub>
- [x] Freebies are referenced in commit titles              <sub>or there are no freebies in this PR</sub>
- [x] This PR is blocked by previous PR in the chain        <sub>or this PR is not chained to another PR</sub>
- [x] Added `chain` label to the blocking PR                <sub>or this PR is not chained to another PR</sub>

Author (upgrading)

- [x] Documented upgrading of deployments in UPGRADING.rst  <sub>or this PR does not require upgrading</sub>
- [x] Added `u` tag to commit title                         <sub>or this PR does not require upgrading</sub>
- [x] Added `upgrade` label to PR                           <sub>or this PR does not require upgrading</sub>
- [x] Added announcement to PR description                  <sub>or this PR does not require announcement</sub>

Author (requirements, before every review)

- [x] Ran `make requirements_update`                        <sub>or this PR leaves requirements*.txt, common.mk and Makefile untouched</sub>
- [x] Added `R` tag to commit title                         <sub>or this PR leaves requirements*.txt untouched</sub>
- [x] Added `reqs` label to PR                              <sub>or this PR leaves requirements*.txt untouched</sub>

Author (before every review)

- [x] `make integration_test` passes in personal deployment <sub>or this PR does not touch functionality that could break the IT</sub>
- [x] Rebased branch on `develop`, squashed old fixups

Primary reviewer (after approval)

- [x] Commented in issue about demo expectations            <sub>or labelled issue as `no demo`</sub>
- [x] Decided if PR can be labeled `no sandbox`
- [x] PR title is appropriate as title of merge commit
- [x] Moved ticket to Approved column
- [x] Assigned PR to an operator

Operator (before pushing merge the commit)

- [x] Checked `reindex` label and `r` commit title tag
- [x] Checked that demo expectations are clear              <sub>or issue is labeled as `no demo`</sub>
- [x] Rebased and squashed branch
- [x] Sanity-checked history
- [x] Pushed PR branch to Github
- [x] Branch pushed to Gitlab and added `sandbox` label     <sub>or PR is labeled `no sandbox`</sub>
- [x] Build passed in sandbox                               <sub>or PR is labeled `no sandbox`</sub>
- [x] Started reindex in `sandbox`                          <sub>or this PR does not require reindexing `sandbox`</sub>
- [x] Checked for failures in `sandbox`                     <sub>or this PR does not require reindexing `sandbox`</sub>
- [x] Added PR reference to merge commit title
- [x] Collected commit title tags in merge commit title
- [x] Moved linked issue to Merged column
- [ ] Pushed merge commit to Github

Operator (after pushing the merge commit)

- [ ] Made announcement requested by author                 <sub>or PR description does not contain an announcement</sub>
- [ ] Moved freebies to Merged column                       <sub>or there are no freebies in this PR</sub> 
- [ ] Shortened the PR chain                                <sub>or this PR is not the base of another PR</sub>
- [ ] Verified that `N reviews` labelling is accurate
- [ ] Pushed merge commit to Gitlab                         <sub>or this changes can be pushed later, together with another PR</sub>
- [ ] Deleted PR branch from Github and Gitlab

Operator (reindex) 

- [ ] Started reindex in `dev`                              <sub>or this PR does not require reindexing or does not target `dev`</sub>
- [ ] Checked for failures in `dev`                         <sub>or this PR does not require reindexing or does not target `dev`</sub>
- [ ] Started reindex in `prod`                             <sub>or this PR does not require reindexing or does not target `prod`</sub>
- [ ] Checked for failures in `prod`                        <sub>or this PR does not require reindexing or does not target `prod`</sub>

Operator

- [ ] Unassigned PR

### CODE REVIEW

**Comment by jessebrennan** [comment]
the commit titled "Check layer zip for duplicates (928bb0d)" causes the layer generation script without the fix ("Fix: Lambda layer zip file exceeded deployment package size limit (#3329)" 9ce3fd8). Here's what the failure looks like:
```
$ python scripts/stage_layer.py
2021-08-17 15:20:40,597 INFO    MainThread: Using dependencies layer package at s3://edu-ucsc-gi-singlecell-azul-config-dev.us-east-1/azul/lambda_layer/8a722d4724e031a3a5865e66116ce227eaa76843.f01861499c0b2edd2441de0c261e3eb3d5d4b658.zip.
2021-08-17 15:20:40,597 INFO    MainThread: Checking if layer package needs updating ...
2021-08-17 15:20:40,608 INFO    MainThread: Found credentials in shared credentials file: ~/.aws/credentials
2021-08-17 15:20:41,083 INFO    MainThread: Staging layer package ...
2021-08-17 15:20:41,084 INFO    MainThread: Running ['chalice', 'package', PosixPath('/home/jesse/gi/azul.2/lambdas/layer/.chalice/terraform')]
Creating deployment package.
/home/jesse/.pyenv/versions/3.8.3/lib/python3.8/zipfile.py:1516: UserWarning: Duplicate name: 'jsonobject/exceptions.py'
  return self._open_to_write(zinfo, force_zip64=force_zip64)
/home/jesse/.pyenv/versions/3.8.3/lib/python3.8/zipfile.py:1516: UserWarning: Duplicate name: 'jsonobject/utils.cpython-38-x86_64-linux-gnu.so'
  return self._open_to_write(zinfo, force_zip64=force_zip64)
/home/jesse/.pyenv/versions/3.8.3/lib/python3.8/zipfile.py:1516: UserWarning: Duplicate name: 'jsonobject/base.cpython-38-x86_64-linux-gnu.so'
  return self._open_to_write(zinfo, force_zip64=force_zip64)
/home/jesse/.pyenv/versions/3.8.3/lib/python3.8/zipfile.py:1516: UserWarning: Duplicate name: 'jsonobject/containers.cpython-38-x86_64-linux-gnu.so'
  return self._open_to_write(zinfo, force_zip64=force_zip64)
/home/jesse/.pyenv/versions/3.8.3/lib/python3.8/zipfile.py:1516: UserWarning: Duplicate name: 'jsonobject/properties.cpython-38-x86_64-linux-gnu.so'
  return self._open_to_write(zinfo, force_zip64=force_zip64)
/home/jesse/.pyenv/versions/3.8.3/lib/python3.8/zipfile.py:1516: UserWarning: Duplicate name: 'jsonobject/base_properties.cpython-38-x86_64-linux-gnu.so'
  return self._open_to_write(zinfo, force_zip64=force_zip64)
/home/jesse/.pyenv/versions/3.8.3/lib/python3.8/zipfile.py:1516: UserWarning: Duplicate name: 'jsonobject/api.cpython-38-x86_64-linux-gnu.so'
  return self._open_to_write(zinfo, force_zip64=force_zip64)
/home/jesse/.pyenv/versions/3.8.3/lib/python3.8/zipfile.py:1516: UserWarning: Duplicate name: 'jsonobject/__init__.py'
  return self._open_to_write(zinfo, force_zip64=force_zip64)
/home/jesse/.pyenv/versions/3.8.3/lib/python3.8/zipfile.py:1516: UserWarning: Duplicate name: 'jsonobject-0.9.9.dist-info/WHEEL'
  return self._open_to_write(zinfo, force_zip64=force_zip64)
/home/jesse/.pyenv/versions/3.8.3/lib/python3.8/zipfile.py:1516: UserWarning: Duplicate name: 'jsonobject-0.9.9.dist-info/top_level.txt'
  return self._open_to_write(zinfo, force_zip64=force_zip64)
/home/jesse/.pyenv/versions/3.8.3/lib/python3.8/zipfile.py:1516: UserWarning: Duplicate name: 'jsonobject-0.9.9.dist-info/LICENSE'
  return self._open_to_write(zinfo, force_zip64=force_zip64)
/home/jesse/.pyenv/versions/3.8.3/lib/python3.8/zipfile.py:1516: UserWarning: Duplicate name: 'jsonobject-0.9.9.dist-info/RECORD'
  return self._open_to_write(zinfo, force_zip64=force_zip64)
/home/jesse/.pyenv/versions/3.8.3/lib/python3.8/zipfile.py:1516: UserWarning: Duplicate name: 'jsonobject-0.9.9.dist-info/METADATA'
  return self._open_to_write(zinfo, force_zip64=force_zip64)
2021-08-17 15:25:55,712 INFO    MainThread: Packaging /home/jesse/gi/azul.2/lambdas/layer/.chalice/terraform/layer.zip
/home/jesse/.pyenv/versions/3.8.3/lib/python3.8/zipfile.py:1516: UserWarning: Duplicate name: 'python/jsonobject/exceptions.py'
  return self._open_to_write(zinfo, force_zip64=force_zip64)
/home/jesse/.pyenv/versions/3.8.3/lib/python3.8/zipfile.py:1516: UserWarning: Duplicate name: 'python/jsonobject/utils.cpython-38-x86_64-linux-gnu.so'
  return self._open_to_write(zinfo, force_zip64=force_zip64)
/home/jesse/.pyenv/versions/3.8.3/lib/python3.8/zipfile.py:1516: UserWarning: Duplicate name: 'python/jsonobject/base.cpython-38-x86_64-linux-gnu.so'
  return self._open_to_write(zinfo, force_zip64=force_zip64)
/home/jesse/.pyenv/versions/3.8.3/lib/python3.8/zipfile.py:1516: UserWarning: Duplicate name: 'python/jsonobject/containers.cpython-38-x86_64-linux-gnu.so'
  return self._open_to_write(zinfo, force_zip64=force_zip64)
/home/jesse/.pyenv/versions/3.8.3/lib/python3.8/zipfile.py:1516: UserWarning: Duplicate name: 'python/jsonobject/properties.cpython-38-x86_64-linux-gnu.so'
  return self._open_to_write(zinfo, force_zip64=force_zip64)
/home/jesse/.pyenv/versions/3.8.3/lib/python3.8/zipfile.py:1516: UserWarning: Duplicate name: 'python/jsonobject/base_properties.cpython-38-x86_64-linux-gnu.so'
  return self._open_to_write(zinfo, force_zip64=force_zip64)
/home/jesse/.pyenv/versions/3.8.3/lib/python3.8/zipfile.py:1516: UserWarning: Duplicate name: 'python/jsonobject/api.cpython-38-x86_64-linux-gnu.so'
  return self._open_to_write(zinfo, force_zip64=force_zip64)
/home/jesse/.pyenv/versions/3.8.3/lib/python3.8/zipfile.py:1516: UserWarning: Duplicate name: 'python/jsonobject/__init__.py'
  return self._open_to_write(zinfo, force_zip64=force_zip64)
/home/jesse/.pyenv/versions/3.8.3/lib/python3.8/zipfile.py:1516: UserWarning: Duplicate name: 'python/jsonobject-0.9.9.dist-info/WHEEL'
  return self._open_to_write(zinfo, force_zip64=force_zip64)
/home/jesse/.pyenv/versions/3.8.3/lib/python3.8/zipfile.py:1516: UserWarning: Duplicate name: 'python/jsonobject-0.9.9.dist-info/top_level.txt'
  return self._open_to_write(zinfo, force_zip64=force_zip64)
/home/jesse/.pyenv/versions/3.8.3/lib/python3.8/zipfile.py:1516: UserWarning: Duplicate name: 'python/jsonobject-0.9.9.dist-info/LICENSE'
  return self._open_to_write(zinfo, force_zip64=force_zip64)
/home/jesse/.pyenv/versions/3.8.3/lib/python3.8/zipfile.py:1516: UserWarning: Duplicate name: 'python/jsonobject-0.9.9.dist-info/RECORD'
  return self._open_to_write(zinfo, force_zip64=force_zip64)
/home/jesse/.pyenv/versions/3.8.3/lib/python3.8/zipfile.py:1516: UserWarning: Duplicate name: 'python/jsonobject-0.9.9.dist-info/METADATA'
  return self._open_to_write(zinfo, force_zip64=force_zip64)
Traceback (most recent call last):
  File "scripts/stage_layer.py", line 36, in <module>
    main(sys.argv[1:])
  File "scripts/stage_layer.py", line 31, in main
    layer.update_layer(force=options.force)
  File "/home/jesse/gi/azul.2/src/azul/lambda_layer.py", line 60, in update_layer
    self._validate_layer(layer_zip)
  File "/home/jesse/gi/azul.2/src/azul/lambda_layer.py", line 93, in _validate_layer
    assert not duplicates, duplicates
AssertionError: {'python/jsonobject/exceptions.py': [<ZipInfo filename='python/jsonobject/exceptions.py' compress_type=deflate filemode='-rw-rw-r--' file_size=239 compress_size=153>, <ZipInfo filename='python/jsonobject/exceptions.py' compress_type=deflate filemode='-rw-r--r--' file_size=239 compress_size=153>], 'python/jsonobject/utils.cpython-38-x86_64-linux-gnu.so': [<ZipInfo filename='python/jsonobject/utils.cpython-38-x86_64-linux-gnu.so' compress_type=deflate filemode='-rw-rw-r--' file_size=435608 compress_size=132511>, <ZipInfo filename='python/jsonobject/utils.cpython-38-x86_64-linux-gnu.so' compress_type=deflate filemode='-rwxr-xr-x' file_size=486600 compress_size=144108>], 'python/jsonobject/base.cpython-38-x86_64-linux-gnu.so': [<ZipInfo filename='python/jsonobject/base.cpython-38-x86_64-linux-gnu.so' compress_type=deflate filemode='-rw-rw-r--' file_size=1747592 compress_size=501655>, <ZipInfo filename='python/jsonobject/base.cpython-38-x86_64-linux-gnu.so' compress_type=deflate filemode='-rwxr-xr-x' file_size=2036784 compress_size=482548>], 'python/jsonobject/containers.cpython-38-x86_64-linux-gnu.so': [<ZipInfo filename='python/jsonobject/containers.cpython-38-x86_64-linux-gnu.so' compress_type=deflate filemode='-rw-rw-r--' file_size=1635248 compress_size=461836>, <ZipInfo filename='python/jsonobject/containers.cpython-38-x86_64-linux-gnu.so' compress_type=deflate filemode='-rwxr-xr-x' file_size=1722952 compress_size=437136>], 'python/jsonobject/properties.cpython-38-x86_64-linux-gnu.so': [<ZipInfo filename='python/jsonobject/properties.cpython-38-x86_64-linux-gnu.so' compress_type=deflate filemode='-rw-rw-r--' file_size=825840 compress_size=238021>, <ZipInfo filename='python/jsonobject/properties.cpython-38-x86_64-linux-gnu.so' compress_type=deflate filemode='-rwxr-xr-x' file_size=935328 compress_size=230311>], 'python/jsonobject/base_properties.cpython-38-x86_64-linux-gnu.so': [<ZipInfo filename='python/jsonobject/base_properties.cpython-38-x86_64-linux-gnu.so' compress_type=deflate filemode='-rw-rw-r--' file_size=1364792 compress_size=386872>, <ZipInfo filename='python/jsonobject/base_properties.cpython-38-x86_64-linux-gnu.so' compress_type=deflate filemode='-rwxr-xr-x' file_size=1403344 compress_size=363349>], 'python/jsonobject/api.cpython-38-x86_64-linux-gnu.so': [<ZipInfo filename='python/jsonobject/api.cpython-38-x86_64-linux-gnu.so' compress_type=deflate filemode='-rw-rw-r--' file_size=306224 compress_size=94191>, <ZipInfo filename='python/jsonobject/api.cpython-38-x86_64-linux-gnu.so' compress_type=deflate filemode='-rwxr-xr-x' file_size=418688 compress_size=100555>], 'python/jsonobject/__init__.py': [<ZipInfo filename='python/jsonobject/__init__.py' compress_type=deflate filemode='-rw-rw-r--' file_size=840 compress_size=269>, <ZipInfo filename='python/jsonobject/__init__.py' compress_type=deflate filemode='-rw-r--r--' file_size=840 compress_size=269>], 'python/jsonobject-0.9.9.dist-info/WHEEL': [<ZipInfo filename='python/jsonobject-0.9.9.dist-info/WHEEL' compress_type=deflate filemode='-rw-rw-r--' file_size=103 compress_size=99>, <ZipInfo filename='python/jsonobject-0.9.9.dist-info/WHEEL' compress_type=deflate filemode='-rw-r--r--' file_size=103 compress_size=99>], 'python/jsonobject-0.9.9.dist-info/top_level.txt': [<ZipInfo filename='python/jsonobject-0.9.9.dist-info/top_level.txt' compress_type=deflate filemode='-rw-rw-r--' file_size=11 compress_size=13>, <ZipInfo filename='python/jsonobject-0.9.9.dist-info/top_level.txt' compress_type=deflate filemode='-rw-r--r--' file_size=11 compress_size=13>], 'python/jsonobject-0.9.9.dist-info/LICENSE': [<ZipInfo filename='python/jsonobject-0.9.9.dist-info/LICENSE' compress_type=deflate filemode='-rw-rw-r--' file_size=2844 compress_size=1360>, <ZipInfo filename='python/jsonobject-0.9.9.dist-info/LICENSE' compress_type=deflate filemode='-rw-r--r--' file_size=2844 compress_size=1360>], 'python/jsonobject-0.9.9.dist-info/RECORD': [<ZipInfo filename='python/jsonobject-0.9.9.dist-info/RECORD' compress_type=deflate filemode='-rw-rw-r--' file_size=1217 compress_size=661>, <ZipInfo filename='python/jsonobject-0.9.9.dist-info/RECORD' compress_type=deflate filemode='-rw-rw-r--' file_size=1217 compress_size=665>], 'python/jsonobject-0.9.9.dist-info/METADATA': [<ZipInfo filename='python/jsonobject-0.9.9.dist-info/METADATA' compress_type=deflate filemode='-rw-rw-r--' file_size=14115 compress_size=5007>, <ZipInfo filename='python/jsonobject-0.9.9.dist-info/METADATA' compress_type=deflate filemode='-rw-r--r--' file_size=14115 compress_size=5007>]}
```

File: src/azul/lambda_layer.py [hunk 1]
```diff
@@ -23,8 +32,47 @@
...
+    name: str
+    version: str
+    git: bool
+
+    wheel_re = re.compile(r'(?P<distribution>[^-]*)-'
+                          r'(?P<version>[^-]*)-'
+                          r'(?:(?P<build_tag>[^-]*)-)?'
+                          r'(?P<python_tag>[^-]*)-'
+                          r'(?P<abi_tag>[^-]*)-'
+                          r'(?P<platform_tag>[^-]*).whl')
```
Comment (danielsotirhos) [inline src/azul/lambda_layer.py:45]:
```suggestion
    wheel_re = re.compile(r'(?P<distribution>[^-]+)-'
                          r'(?P<version>[^-]+)-'
                          r'(?:(?P<build_tag>\d[^-]+)-)?'
                          r'(?P<python_tag>[^-]+)-'
                          r'(?P<abi_tag>[^-]+)-'
                          r'(?P<platform_tag>[^-]+).whl')
```
According to https://www.python.org/dev/peps/pep-0427/#file-name-convention the build tag must start with a number.

File: src/azul/lambda_layer.py [hunk 2]
```diff
@@ -23,8 +32,47 @@
...
+
+    @classmethod
+    def from_pin(cls, line: str):
+        if line.startswith('git+'):
+            line = line[len('git+'):]
+            version, name = line.split('#egg=')
+            return cls(name=name, version=version, git=True)
+        else:
+            name, version = line.split('==')
```
Comment (danielsotirhos) [inline src/azul/lambda_layer.py:54]: Should the other operators be considered (e.g. `>=` or `-=`) https://www.python.org/dev/peps/pep-0440/#version-specifiers or is it guaranteed we'll never anything except `==`?

**Review by danielsotirhos — CHANGES_REQUESTED** [review]

File: src/azul/lambda_layer.py (hunk 2 above)
Comment (jessebrennan) [reply src/azul/lambda_layer.py:54]: We always pin all of our requirements with `==`. Other operators should raise an error.

File: src/azul/lambda_layer.py (hunk 1 above)
Comment (jessebrennan) [reply src/azul/lambda_layer.py:45]: Good catch.

File: lambdas/layer/.gitignore [hunk 3]
```diff
@@ -1,3 +1,4 @@
 /.chalice/*
 !/.chalice/config.json.template.py
 /vendor/*
+requirements.txt
```
Comment (danielsotirhos) [inline lambdas/layer/.gitignore:4]:
```suggestion
requirements.txt

```

**Review by danielsotirhos — CHANGES_REQUESTED** [review]
Sorry I missed this the first review. The rest looks great.

**Review by danielsotirhos — APPROVED** [review]
Approved.

File: src/azul/lambda_layer.py [hunk 4]
```diff
@@ -23,8 +32,47 @@
 log = logging.getLogger(__name__)
 
 
+class Req(NamedTuple):
```
Comment (hannes-ucsc) [inline src/azul/lambda_layer.py:35]: Prefer `attrs` b/c it has `kwonly=`

Comment (hannes-ucsc) [reply src/azul/lambda_layer.py:35]:
```suggestion
class Requirement(NamedTuple):
```

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]
Consider using https://github.com/jwodder/wheel-filename for parsing wheel file names and https://pypi.org/project/requirements-parser/ for `requirement.txt`.

**Comment by jessebrennan** [comment]
> Consider using https://github.com/jwodder/wheel-filename for parsing wheel file names and https://pypi.org/project/requirements-parser/ for requirement.txt.

In stand up, @hannes-ucsc decided to postpone this change in a follow-up ticket: #3342.

**Review by hannes-ucsc — APPROVED** [review]
Approval conditional on tests passing.

**Pull request merged**